File: 2025-11-November-GFG-POTD/November 07 - Weighted Job Scheduling/Solution.java
class Solution {
    public int maxProfit(int[][] jobs) {
        // Sort jobs by end time
        Arrays.sort(jobs, (a, b) -> Integer.compare(a[1], b[1]));
        
        int n = jobs.length;
        int[] dp = new int[n]; // dp[i] stores max profit till job i
        dp[0] = jobs[0][2];
        
        for (int i = 1; i < n; i++) {
            int includeProfit = jobs[i][2];
            
            // Find last non-conflicting job using binary search
            int l = findLastNonConflict(jobs, i);
            if (l != -1) {
                includeProfit += dp[l];
            }
            
            // Exclude current job
            dp[i] = Math.max(dp[i - 1], includeProfit);
        }
        
        return dp[n - 1];
    }
    
    // Binary search: find last job whose end time <= current job's start time
    private int findLastNonConflict(int[][] jobs, int i) {
        int low = 0, high = i - 1;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (jobs[mid][1] <= jobs[i][0]) {
                if (jobs[mid + 1][1] <= jobs[i][0])
                    low = mid + 1;
                else
                    return mid;
            } else {
                high = mid - 1;
            }
        }
        return -1;
    }
}

File: 2025-11-November-GFG-POTD/November 07 - Weighted Job Scheduling/Solution.py
class Solution:
    def maxProfit(self, jobs):
        # Sort jobs by end time
        jobs.sort(key=lambda x: x[1])
        n = len(jobs)
        dp = [0] * n
        dp[0] = jobs[0][2]

        for i in range(1, n):
            include_profit = jobs[i][2]
            
            # Find last non-conflicting job using binary search
            l = self.find_last_non_conflict(jobs, i)
            if l != -1:
                include_profit += dp[l]
            
            dp[i] = max(dp[i - 1], include_profit)
        
        return dp[-1]

    def find_last_non_conflict(self, jobs, i):
        low, high = 0, i - 1
        while low <= high:
            mid = (low + high) // 2
            if jobs[mid][1] <= jobs[i][0]:
                if mid + 1 < i and jobs[mid + 1][1] <= jobs[i][0]:
                    low = mid + 1
                else:
                    return mid
            else:
                high = mid - 1
        return -1
